flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ShowCreateUtil.java

```diff
@@ -0,0 +1,221 @@
+import org.apache.flink.table.catalog.QueryOperationCatalogView;
+import org.apache.flink.table.catalog.ResolvedCatalogBaseTable;
+import org.apache.flink.table.catalog.ResolvedCatalogTable;
+import org.apache.flink.table.catalog.UniqueConstraint;
+import org.apache.flink.table.utils.EncodingUtils;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/** SHOW CREATE statement Util. */
+@Internal
+enum ShowCreateUtil {
+    ;
+
+    static String buildCreateFormattedPrefix(
+            String tableType, boolean isTemporary, ObjectIdentifier identifier) {
+        return String.format(
+                "CREATE %s%s %s (\n",
+                isTemporary ? "TEMPORARY " : "", tableType, identifier.asSerializableString());
+    }
+
+    static Optional<String> extractFormattedPrimaryKey(
+            ResolvedCatalogBaseTable<?> table, String printIndent) {
+        Optional<UniqueConstraint> primaryKey = table.getResolvedSchema().getPrimaryKey();
+        return primaryKey.map(
+                uniqueConstraint -> String.format("%s%s", printIndent, uniqueConstraint));
+    }
+
+    static String getColumnString(Column column) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(EncodingUtils.escapeIdentifier(column.getName()));
+        sb.append(" ");
+        // skip data type for computed column
+        if (column instanceof Column.ComputedColumn) {
+            sb.append(
+                    column.explainExtras()
+                            .orElseThrow(
+                                    () ->
+                                            new TableException(
+                                                    String.format(
+                                                            "Column expression can not be null for computed column '%s'",
+                                                            column.getName()))));
+        } else {
+            sb.append(column.getDataType().getLogicalType().asSerializableString());
+            column.explainExtras()
+                    .ifPresent(
+                            e -> {
+                                sb.append(" ");
+                                sb.append(e);
+                            });
+        }
+        // TODO: Print the column comment until FLINK-18958 is fixed
+        return sb.toString();
+    }
+
+    static String extractFormattedColumns(ResolvedCatalogBaseTable<?> table, String printIndent) {
+        return table.getResolvedSchema().getColumns().stream()
+                .map(column -> String.format("%s%s", printIndent, getColumnString(column)))
+                .collect(Collectors.joining(",\n"));
+    }
+
+    static Optional<String> extractFormattedWatermarkSpecs(
+            ResolvedCatalogBaseTable<?> table, String printIndent) {
+        if (!table.getResolvedSchema().getWatermarkSpecs().isEmpty()) {
```

Comment:
nit: we can invert this condition to exit early and reduce the indentation level:

```
if (table.getResolvedSchema().getWatermarkSpecs().isEmpty()) {
    return Optional.empty();
}

// …
```